File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/DepartEventFactory.java
package com.mapbox.services.android.navigation.v5.internal.navigation;

import com.mapbox.services.android.navigation.v5.internal.location.MetricsLocation;
import com.mapbox.services.android.navigation.v5.navigation.metrics.SessionState;
import com.mapbox.services.android.navigation.v5.routeprogress.MetricsRouteProgress;

import java.util.Date;

class DepartEventFactory {

  private static final int INITIAL_LEG_INDEX = -1;
  private final DepartEventHandler departEventHandler;
  private int currentLegIndex = INITIAL_LEG_INDEX;

  DepartEventFactory(DepartEventHandler departEventHandler) {
    this.departEventHandler = departEventHandler;
  }

  SessionState send(SessionState sessionState, MetricsRouteProgress routeProgress, MetricsLocation location) {
    sessionState = checkResetForNewLeg(sessionState, routeProgress);
    this.currentLegIndex = routeProgress.getLegIndex();
    if (isValidDeparture(sessionState, routeProgress)) {
      return sendToHandler(sessionState, routeProgress, location);
    }
    return sessionState;
  }

  void reset() {
    currentLegIndex = INITIAL_LEG_INDEX;
  }

  private SessionState checkResetForNewLeg(SessionState sessionState, MetricsRouteProgress routeProgress) {
    if (shouldResetDepartureDate(routeProgress)) {
      sessionState = sessionState.toBuilder().startTimestamp(null).build();
    }
    return sessionState;
  }

  private boolean shouldResetDepartureDate(MetricsRouteProgress routeProgress) {
    return currentLegIndex != routeProgress.getLegIndex();
  }

  private boolean isValidDeparture(SessionState sessionState, MetricsRouteProgress routeProgress) {
    return sessionState.startTimestamp() == null && routeProgress.getDistanceTraveled() > 0;
  }

  private SessionState sendToHandler(SessionState sessionState, MetricsRouteProgress routeProgress,
                                     MetricsLocation location) {
    SessionState updatedState = sessionState.toBuilder()
      .startTimestamp(new Date())
      .build();
    departEventHandler.send(updatedState, routeProgress, location);
    return updatedState;
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/DepartEventFactory.kt
package com.mapbox.services.android.navigation.v5.internal.navigation

import com.mapbox.services.android.navigation.v5.internal.location.MetricsLocation
import com.mapbox.services.android.navigation.v5.internal.navigation.metrics.SessionState
import com.mapbox.services.android.navigation.v5.internal.navigation.routeprogress.MetricsRouteProgress
import java.util.Date

internal class DepartEventFactory(private val departEventHandler: DepartEventHandler) {

    companion object {
        private const val INITIAL_LEG_INDEX = -1
    }

    private var currentLegIndex = INITIAL_LEG_INDEX

    fun send(
        sessionState: SessionState,
        routeProgress: MetricsRouteProgress,
        location: MetricsLocation
    ): SessionState {
        val checkedSessionState = checkResetForNewLeg(sessionState, routeProgress)
        this.currentLegIndex = routeProgress.legIndex
        return if (isValidDeparture(checkedSessionState, routeProgress)) {
            sendToHandler(checkedSessionState, routeProgress, location)
        } else {
            checkedSessionState
        }
    }

    fun reset() {
        currentLegIndex = INITIAL_LEG_INDEX
    }

    private fun checkResetForNewLeg(
        sessionState: SessionState,
        routeProgress: MetricsRouteProgress
    ): SessionState {
        if (shouldResetDepartureDate(routeProgress)) {
            sessionState.startTimestamp = null
        }
        return sessionState
    }

    private fun shouldResetDepartureDate(routeProgress: MetricsRouteProgress): Boolean =
        currentLegIndex != routeProgress.legIndex

    private fun isValidDeparture(
        sessionState: SessionState,
        routeProgress: MetricsRouteProgress
    ): Boolean =
        sessionState.startTimestamp == null && routeProgress.distanceTraveled > 0

    private fun sendToHandler(
        sessionState: SessionState,
        routeProgress: MetricsRouteProgress,
        location: MetricsLocation
    ): SessionState {
        sessionState.startTimestamp = Date()
        departEventHandler.send(sessionState, routeProgress, location)
        return sessionState
    }
}
